src/facade/disk_connection_backpressure.h

```diff
@@ -0,0 +1,65 @@
+// Copyright 2025, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <util/fibers/uring_file.h>
+
+#include <deque>
+#include <functional>
+#include <memory>
+#include <string_view>
+#include <system_error>
+
+#include "io/io.h"
+
+namespace facade {
+
+class DiskBackedBackpressureQueue {
+ public:
+  explicit DiskBackedBackpressureQueue(uint32_t conn_id);
+  ~DiskBackedBackpressureQueue();
+
+  std::error_code Init();
+
+  // Check if we can offload bytes to backing file.
+  bool HasEnoughBackingSpaceFor(size_t bytes) const;
+
+  // Total size of internal buffers/structures.
+  size_t TotalInMemoryBytes() const;
+
+  void OffloadToBacking(std::string_view blob);
+
+  // For each item loaded from disk it calls f(item) to consume it.
+  // Reads up to max_queue_load_size_ items on each call
+  void LoadFromBacking(std::function<void(io::MutableBytes)> f);
+
+  // Check if backing file is empty, i.e. backing file has 0 bytes.
+  bool Empty() const;
+
+ private:
+  // File Reader/Writer
+  std::unique_ptr<io::WriteFile> writer_;
+  std::unique_ptr<io::ReadonlyFile> reader_;
+
+  // In memory backed file map
+  struct ItemOffset {
+    size_t offset = 0;
+    size_t total_bytes = 0;
+  };
+
+  std::deque<ItemOffset> offsets_;
```

Comment:
why do we queue of offsets_ ?

Reply:
I think what you suspect is correct. I imagine you have something like this in mind:

[8 bytes + msg] [8 bytes + msg] 

First 8 bytes are the size of the message. Now you can:

Read offset 0..8 -> msg size X
Read msg of size X ->
Read offset 8 + size X -> msg size of Y 
and so on and that way you don't need to track any of the items.

IMO I wanted a simple solution first but if you think it's better to add it here I am happy to do so